pkg/util/converter/prom.go

```diff
@@ -165,35 +167,20 @@ l1Fields:
 			if err != nil {
 				return rspErr(err)
 			}
+			// if we have saved resultBytes we will parse them here
+			// we saved them because when we had them we don't know the resultType
+			if len(resultBytes) > 0 {
+				ji := jsonitere.NewIterator(jsoniter.ParseBytes(jsoniter.ConfigDefault, resultBytes))
+				rsp = readResult(resultType, rsp, ji, opt)
+			}
 		case "result":
-			switch resultType {
-			case "matrix", "vector":
-				rsp = readMatrixOrVectorMulti(iter, resultType, opt)
-				if rsp.Error != nil {
-					return rsp
-				}
-			case "streams":
-				rsp = readStream(iter)
-				if rsp.Error != nil {
-					return rsp
-				}
-			case "string":
-				rsp = readString(iter)
-				if rsp.Error != nil {
-					return rsp
-				}
-			case "scalar":
-				rsp = readScalar(iter)
-				if rsp.Error != nil {
-					return rsp
-				}
-			default:
-				if err = iter.Skip(); err != nil {
-					return rspErr(err)
-				}
-				rsp = backend.DataResponse{
-					Error: fmt.Errorf("unknown result type: %s", resultType),
-				}
+			// for some rare cases resultType is coming after the result.
+			// when that happens we save the bytes and parse them after reading resultType
+			// see: https://github.com/grafana/grafana/issues/64693
+			if resultType != "" {
+				rsp = readResult(resultType, rsp, iter, opt)
+			} else {
+				resultBytes = iter.SkipAndReturnBytes()
```

Comment:
There does appear to be a couple of unlikely edge cases we are not handling here, if resultType is not set at all or set to an empty string then we'll populate resultBytes but never call readResult (because we never reach line 166 due to missing resultType, or because we find an empty resultType before we find result, so we don't hit line 174 or line 181).  If we have an empty resultType *after* (a non-empty) result, then we'll go through into readResult as expected on line 174 and return the unknown result type error.

Reply:
That's absolutely true. Those are extremely edge but valid cases. I will cover those too. Thanks for pointing them.

Reply:
@DanCech Please see the latest state of the PR